rfc/2022/2022-07-20 - Update Build Directory Structure of vets-website.md

```diff
@@ -0,0 +1,61 @@
+
+In the future, the subdirectories of isolated application bundles can be moved to `build/dist` to follow a more standard naming convention. 
+
+### Updating the content build for the new directory structure
+
+The `content-build` repository is responsible for creating the HTML files that are served on the site. By default, the build process in the `content-build` repository uses the `build/generated` path when injecting script tags that reference bundles and assets from `vets-website` deployments. This will need to be modified, since assets could be loaded from `build/generated` or one of the subdirectories for isolated application builds, e.g. `build/generated/vaos`. To better determine which paths in the deployment buckets to load bundles and assets from on each page, a generic script that handles this logic can be created. The build process in the `content-build` repository can then be updated to inject that generic script on each page. This essentially transfers control of determining which assets get loaded to `vets-website`, making it unlikely that production would break for applications migrating to the isolated structure.
+
+
+### Updating CI for the new directory structure
+
+Since there will no longer be a true *full build* once isolated application bundles are stored in their own subdirectory, we'll need to modify the process of building and testing global changes in CI. It won't be possible to build and test every application in one build anymore. One approach that would allow us to build and test multiple builds simultaneously involves Yarn Workspaces. With Yarn workspaces, each isolated application will have its own `package.json` file in the root of its directory, e.g. `src/applications/vaos/package.json`. The `package.json` files can be used to run the commands for each directory in `src/applications` that contains a `package.json`. This means we can use a command like `yarn workspaces foreach` at the root level to run commands like `build`, `test:unit`, `cy:run` (run Cypress tests), etc. The `workspaces foreach` command also has a `--parallel` option that can run the commands in parallel.
+
+## Risks
+
+One risk of this approach is that site wide elements rendered on pages of isolated applications can potentially get out of sync with site wide elements on non-isolated application pages. One example is if the header for the site has a major update that gets released in the daily production deploy. Since isolated applications would be excluded from the daily production deployment, the header update would not be applied to them at the time of the daily deployment, meaning some applications on the site would be using the new header, while others would be using the old header. This isn't ideal for consistency of the site, as we want all pages on the site to use the same site wide elements.
```

Comment:
Another risk is a performance hit. Currently each global asset is only deployed once, and that single asset is referenced by all apps which need it. The first time the browser loads the asset, it caches the asset locally. If the user loads other apps, the browser uses the cached asset instead of fetching it from the S3 bucket.

If each app deploys its own copy of global assets, the browser wouldn't be able to reliably cache those assets. Instead the global assets would be reloaded when each app loads. In practice I don't know how much of a performance hit this would represent. It depends on how many assets are currently shared between apps, how large the assets are, and how often users load multiple apps.

Reply:
Good point, I'll add that to this section That's something we'd need to test to see how much of a performance impact it would have. If it ends up having a big performance impact, we'd need to figure out a way to mitigate that

Reply:
@theodur Can we spin up a ticket to isolate global assets into their own entrypoint? That would address this issue, esp. since the header/footer is a react component now.

If we do start adding scripts into the DOM I would also want to try and move the site from HTTP v1 to HTTP v2, which will give us a performance boost.

Reply:
@timwright12 That's a good idea, that should work for global assets like the header, which function independently of applications. We would still run into issues with global bundles that apps depend on though, like `web-components.entry.js`, `vendor.entry.js`, etc. because the contents of those bundles could vary between isolated apps, since they’d be deployed at different times. An isolated app build could pick up changes in a web component update when merged to `main` that other apps haven’t picked up yet if they haven’t been merged to `main` and deployed recently, making it tough to have all apps use the same version of those bundles if they're built and deployed at different points in time on the `main` branch. 

One way around that might be to use chunkhashes or contenthashes in the filename output in the Webpack config. Apps would still deploy a copy of the global bundles that were generated in their build to their subdirectories, but if a file with the same name (which would include the hash) exists outside their subdirectory at the root level, they would use that file instead. This way, apps could still rely on the same cached global bundle files, unless their deployment was built relying on global bundles changes that haven't been deployed yet.

> If we do start adding scripts into the DOM I would also want to try and move the site from HTTP v1 to HTTP v2, which will give us a performance boost.

That would be nice to move to HTTP/2, especially for things like multiplexing, which could help with the changes we'd be making. I can make a ticket for future discovery work on that

Reply:
re: web-components.entry.js, vendor.entry.js

Maybe but, I don't think we should get hung up on files like that which really are never updated

Reply:
It's probably ugly, but, building on Theo's mention of hashes in webpack filenames, what if we let every build create its own versions of every file, but always with hashes, and everything just got uploaded into the same global directory? If multiple builds uploaded the same file with the same contents, the file would have the same hash since the hash is based on the contents. I that case the file would just get overwritten multiple times which is fine. If multiple builds uploaded different versions of the same file, they would have different hashes, and each build app would use its own version. This would allow all apps to share a single deploy directory since the hash provides versioning.

Reply:
@timwright12

> Maybe but, I don't think we should get hung up on files like that which really are never updated

Yeah, in reality those bundles don't get updated as often, especially `vendor.entry.js`. The web components bundle probably gets updated more frequently, although not often with breaking changes. It's the rare instances when they do get updated with breaking changes that I'm worried will come back to haunt us if we don't iron things out now

@alexandec
That's a good point 👀 All builds could deploy their hashed versions to the same directory in the bucket, which would ensure apps have the bundles/assets they need to function correctly, while also being able to use the cached assets in the browser. I like that because it would still allow apps to share global bundles, and even they needed a different version, they would deploy it themselves, which future isolated app deploys could use if they also needed them.

The main challenge would probably be keeping the deployment bucket clean with all the potential different versions of the global bundles. If all apps are still included in the daily production deploy, that deploy would wipe the slate clean for all versions of global bundles, essentially cleaning the deployment bucket. The only thing is there'd still be potential race conditions with isolated app deploys. If isolated apps aren't part of the daily production deploy, we'd need to figure out a way of keeping the deployment bucket clean with versions of global assets that are actually being used by apps. I'm also wondering if there'd be issues with concurrent isolated app deploys potentially writing to the same hash file in the S3 bucket 🤔. Not sure if that would cause flakiness with the deploys.

I'm going to add that into the `Alternatives` section, because that itself is another solution that could work.

Reply:
> we'd need to figure out a way of keeping the deployment bucket clean with versions of global assets that are actually being used by apps

That's true, otherwise you'd end up with an endless number of different versions of very global asset, all with different hashes, most of them unused. One solution is a cleanup script which would check which assets are referenced by other files in the bucket, then delete all unreferenced assets. It's similar to the garbage collection compilers do, deleting unreferenced objects. The script could run daily at 3am when no deploys are happening.

> I'm also wondering if there'd be issues with concurrent isolated app deploys potentially writing to the same hash file in the S3 bucket 🤔. Not sure if that would cause flakiness with the deploys.

I think `s3 sync` would take care of that since it would detect that the file is exactly the same (name and contents) and not overwrite it.

Reply:
That's a good idea about having a cleanup script. That would definitely help keep the bucket from getting populated with unused assets. Having it run early in the morning like you mentioned shouldn't interfere with anything

> I think s3 sync would take care of that since it would detect that the file is exactly the same (name and contents) and not overwrite it.

We would probably need to update our deploy script, because I think files that haven't changed still get synced to the bucket in deployments

Reply:
One thing that came to mind with the approach of uploading hashed bundles to the same directory in the bucket is how would we know which version of bundles like `vendor.entry.js`, `web-components.entry.js`, etc. to inject on application pages. With potentially multiple versions existing in the deployment bucket, we'd need to know which one was used for each application's build in order to dynamically determine which hashed bundles to inject on application pages when they're loaded in the browser

Reply:
That's a good point, I was just thinking about that since I was looking at how `content-build` references those bundles in https://github.com/department-of-veterans-affairs/content-build/blob/main/src/site/includes/header.html (and the corresponding script [process-entry-names.js](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/site/stages/build/plugins/modify-dom/process-entry-names.js)). Instead of letting content-build handle this, I think we'd give webpack control to add script tags to the entry page for each app during the vets-website build process. The script tags would reference the bundles with the correct hashes for that app. It's similar to how we [generate the scaffold pages ](https://github.com/department-of-veterans-affairs/vets-website/blob/main/config/webpack.config.js#L204) currently using `html-webpack-plugin`.

This would have the side benefit of bringing more app build stuff out of content-build templates/scripts, and into vets-website. Possibly also the scaffold pages and the app entry pages could become the same thing. Not sure if I'm thinking of all the edge cases here, but maybe it could work.

Reply:
That's a good idea, I'm glad you brought that up because I was also thinking about generating the app pages in Webpack, just didn't know if it sounded crazy in my head haha. I totally with agree that. It would give us more control over which assets are loaded on app entry pages. That would also remove the need for the dynamic script on app pages, since vets-website would be able to add script tags to the pages at build time.

Also, looking through `process-entry-names.js` reminded me about the teamsite pages. Since those pages [hardcode](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/site/stages/build/plugins/modify-dom/process-entry-names.js#L11-L13) some of the global bundles, hashing those filenames might cause some issues